File: src/main/java/com/yupi/airouter/service/RateLimitService.java
/**
 * 限流服务接口
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
package com.yupi.airouter.service;

import java.time.Duration;

/**
 * 限流服务
 * 提供基于 Redisson 的限流能力
 */
public interface RateLimitService {

    /**
     * 尝试获取限流许可
     *
     * @param key      限流键（如 API Key 或 IP）
     * @param limit    限制次数
     * @param duration 时间窗口
     * @return true 表示允许访问，false 表示被限流
     */
    boolean tryAcquire(String key, int limit, Duration duration);

    /**
     * 获取当前可用许可数
     *
     * @param key 限流键
     * @return 当前可用的许可数
     */
    long getAvailablePermits(String key);

    /**
     * 检查 API Key 是否被限流
     *
     * @param apiKey API Key 值
     * @param limit  限制次数（每秒）
     * @return true 表示允许访问，false 表示被限流
     */
    boolean checkApiKeyRateLimit(String apiKey, int limit);

    /**
     * 检查 IP 是否被限流
     *
     * @param ip    客户端 IP
     * @param limit 限制次数（每秒）
     * @return true 表示允许访问，false 表示被限流
     */
    boolean checkIpRateLimit(String ip, int limit);
}

File: python-backend/app/services/rate_limit_service.py
"""Rate limit service based on Redis fixed window."""

from __future__ import annotations

from redis.asyncio import Redis

from app.core.constants import RATE_LIMIT_API_KEY_PREFIX, RATE_LIMIT_IP_PREFIX


class RateLimitService:
    def __init__(self, redis: Redis) -> None:
        self.redis = redis

    async def try_acquire(self, key: str, limit: int, window_seconds: int) -> bool:
        current = await self.redis.incr(key)
        if current == 1:
            await self.redis.expire(key, window_seconds)
        return int(current) <= limit

    async def get_available_permits(self, key: str, limit: int) -> int:
        current = await self.redis.get(key)
        used = int(current) if current else 0
        remain = limit - used
        return remain if remain > 0 else 0

    async def check_api_key_rate_limit(self, api_key: str, limit: int) -> bool:
        key = f"{RATE_LIMIT_API_KEY_PREFIX}{api_key}"
        return await self.try_acquire(key, limit, 1)

    async def check_ip_rate_limit(self, ip: str, limit: int) -> bool:
        key = f"{RATE_LIMIT_IP_PREFIX}{ip}"
        return await self.try_acquire(key, limit, 1)
